### PR TITLE
ENH: added new instrument, sw_ace

### PR DESCRIPTION
**Pull request by aburrell**
Added new instrument, sw_ace, which handles the four different real-time data sets available from ACE.  This differs from the science-level ACE data in its availability, resolution, and data products.

This branch has been tested locally and all ACE related tests pass.  New unit tests have not been added, as no special functions were created.  Method routines deal with assigning acknowledgements and references.

To try this out to see how it works well:
```
import datetime as dt
import pysat
import pysatSpaceWeather as pysat_sw

stime = dt.datetime(2009,1,1)
amag = pysat.Instrument(inst_module=pysat_sw.instruments.sw_ace, tag='hist', sat_id='mag')
amag.download(start=stime, stop=stime)
amag.load(date=stime)
print(amag.data.iloc[:3])                                              
                          jd    sec  status  ...  bt_gsm  lat_gsm  lon_gsm
2009-01-01 00:00:00  54832.0    0.0     0.0  ...     4.5     23.1    142.7
2009-01-01 00:01:00  54832.0   60.0     0.0  ...     4.5     25.8    147.3
2009-01-01 00:02:00  54832.0  120.0     0.0  ...     4.5     21.8    154.8

[3 rows x 9 columns]
```

And another example uses the real-time data file:
```
sis = pysat.Instrument(inst_module=pysat_sw.instruments.sw_ace, tag='rt', sat_id='sis')
sis.download(start=dt.datetime.now(), stop=dt.datetime.now())
sis.load(date=dt.datetime.now())
print(sis.data.iloc[:3])                                               
                          jd      sec  ...  status_30  int_pflux_30MeV
2020-08-19 19:05:00  59080.0  68700.0  ...        0.0             1.74
2020-08-19 19:10:00  59080.0  69000.0  ...        0.0             1.72
2020-08-19 19:15:00  59080.0  69300.0  ...        0.0             1.70

[3 rows x 6 columns]
```

### CODE REVIEW

File: pysatSpaceWeather/instruments/sw_ace.py [hunk 1]
```diff
@@ -0,0 +1,463 @@
...
+
+Properties
+----------
+platform
+    'sw'
+name
+    'ace'
+tag
+    - 'rt' Real-time data from the Space Weather Prediction Center (SWPC)
+    - 'hist' Historical data from the SWPC
```
Comment (rstoneback) [inline pysatSpaceWeather/instruments/sw_ace.py:12]:
```suggestion
    - 'historic' Historical data from the SWPC
```

File: pysatSpaceWeather/instruments/sw_ace.py [hunk 2]
```diff
@@ -0,0 +1,463 @@
...
+"""Supports ACE Magnetometer data
+
+Properties
+----------
+platform
+    'sw'
+name
+    'ace'
+tag
+    - 'rt' Real-time data from the Space Weather Prediction Center (SWPC)
```
Comment (rstoneback) [inline pysatSpaceWeather/instruments/sw_ace.py:11]:
```suggestion
    - 'real-time' Real-time data from the Space Weather Prediction Center (SWPC)
```

File: pysatSpaceWeather/instruments/sw_ace.py (hunk 1 above)
Comment (aburrell) [reply pysatSpaceWeather/instruments/sw_ace.py]: Changed throughout file in last commit.

File: pysatSpaceWeather/instruments/sw_ace.py (hunk 2 above)
Comment (aburrell) [reply pysatSpaceWeather/instruments/sw_ace.py]: Changes to unhyphenated version throughout file in last commit.

**Comment by aburrell** [comment]
@jklenzing this is now ready for review.

**Comment by rstoneback** [comment]
Review in progress. If you instantiate
`pysat.Instrument('sw', 'ace')` without a tag, the error isn't very clear that the user is missing a tag.

```
KeyError: ''
> /Users/russellstoneback/Code/pysatSpaceWeather/pysatSpaceWeather/instruments/methods/ace.py(68)references()
     64                             'Composition Explorer. Space Science Reviews ',
     65                             '86, 357–408 (1998). ',
     66                             'https://doi.org/10.1023/A:1005027929871'])}
     67 
---> 68     return refs[iname]

```

FYI, branch `registry_update` over in pysat can make review a little easier.

```
In [45]: pysat.utils.registry.register_by_module(pysatSpaceWeather.instruments)                                                                                                       

In [46]: pysat.user_modules                                                                                                                                                            
Out[46]: 
{'sw': {'dst': 'pysatSpaceWeather.instruments.sw_dst',
  'f107': 'pysatSpaceWeather.instruments.sw_f107',
  'kp': 'pysatSpaceWeather.instruments.sw_kp',
  'ace': 'pysatSpaceWeather.instruments.sw_ace'}}

In [47]: inst = pysat.Instrument('sw', 'ace')                                                                                                                                          
pysat WARNING: Unable to find any files that match the supplied template. If you have the necessary files please check pysat settings and file locations (e.g. pysat.pysat_dir).
```

**Comment by aburrell** [comment]
> If you instantiate
pysat.Instrument('sw', 'ace') without a tag, the error isn't very clear that the user is missing a tag.

Great point.  This is going to be an issue for most instruments.  We should probably have a `test_input` routine in the Instrument class that ensures the specified tag and sat_id are correct and call it by default from `__init__` in all pysat Instruments.

File: pysatSpaceWeather/instruments/sw_ace.py [hunk 3]
```diff
@@ -0,0 +1,463 @@
...
+                times.append(dt.datetime.strptime(' '.join(split_line[:4]),
+                                                  '%Y %m %d %H%M'))
+                for i, col in enumerate(data_cols[sat_id]):
+                    # Convert to a number and save
+                    #
+                    # Output is saved as a float, so don't bother to
+                    # differentiate between int and float
+                    data_dict[col].append(float(split_line[4+i]))
+            else:
+                if len(split_line) > 0:
```
Comment (rstoneback) [inline pysatSpaceWeather/instruments/sw_ace.py:385]:
```
In [11]: ace = pysat.Instrument('sw', 'ace', tag='historic', sat_id='mag')                                                                                                             
pysat WARNING: Unable to find any files that match the supplied template. If you have the necessary files please check pysat settings and file locations (e.g. pysat.pysat_dir).

In [12]: ace.download(dt.datetime(2001, 1, 1), dt.datetime(2002, 1, 1))                                                                                                                
---------------------------------------------------------------------------
OSError                                   Traceback (most recent call last)
<ipython-input-12-f95df3dfb9a7> in <module>
----> 1 ace.download(dt.datetime(2001, 1, 1), dt.datetime(2002, 1, 1))

~/Code/pysat/pysat/_instrument.py in download(self, start, stop, freq, user, password, date_array, **kwargs)
   1765 
   1766         if user is None:
-> 1767             self._download_rtn(date_array,
   1768                                tag=self.tag,
   1769                                sat_id=self.sat_id,

~/Code/pysatSpaceWeather/pysatSpaceWeather/instruments/sw_ace.py in download(date_array, tag, sat_id, data_path)
    384             else:
    385                 if len(split_line) > 0:
--> 386                     raise IOError(''.join(['unexpected line encoutered in ',
    387                                            furl, ":\n", raw_line]))
    388 

OSError: unexpected line encoutered in https://sohoftp.nascom.nasa.gov/sdb/ace/daily/20010101_ace_mag_1m.txt:
<html><head>
```

Comment (rstoneback) [reply pysatSpaceWeather/instruments/sw_ace.py:385]:
```
In [14]: ace = pysat.Instrument('sw', 'ace', tag='historic', sat_id='swepam')                                                                                                          
pysat WARNING: Unable to find any files that match the supplied template. If you have the necessary files please check pysat settings and file locations (e.g. pysat.pysat_dir).

In [15]: ace.download(dt.datetime(2001, 1, 1), dt.datetime(2002, 1, 1))                                                                                                                
---------------------------------------------------------------------------
OSError                                   Traceback (most recent call last)
<ipython-input-15-f95df3dfb9a7> in <module>
----> 1 ace.download(dt.datetime(2001, 1, 1), dt.datetime(2002, 1, 1))
OSError: unexpected line encoutered in https://sohoftp.nascom.nasa.gov/sdb/ace/daily/20010101_ace_swepam_1m.txt:
<html><head>
```

Comment (rstoneback) [reply pysatSpaceWeather/instruments/sw_ace.py:385]: Same for the other instruments. I'm running python 3.8 if that is significant.

**Comment by rstoneback** [comment]
All of the realtime instruments worked for me, download, and load. Also checked `mag` specifically for today's data.

```
In [53]: ace.load(date=ace.today())                                                                                                                                 

In [54]: ace.data                                                                                                                                                   
Out[54]: 
                          jd      sec  status  bx_gsm  by_gsm  bz_gsm  bt_gsm  lat_gsm  lon_gsm
2020-08-28 18:24:00  59089.0  66240.0     0.0    -4.7     4.5    -0.0     6.5     -0.0    136.1
2020-08-28 18:25:00  59089.0  66300.0     0.0    -4.4     4.5    -0.4     6.3     -3.8    134.5
2020-08-28 18:26:00  59089.0  66360.0     0.0    -4.3     4.6     0.2     6.3      1.5    133.3
2020-08-28 18:27:00  59089.0  66420.0     0.0    -4.2     4.7     0.5     6.3      4.9    131.9
2020-08-28 18:28:00  59089.0  66480.0     0.0    -4.6     3.9     2.0     6.4     17.8    139.6
...                      ...      ...     ...     ...     ...     ...     ...      ...      ...
2020-08-28 20:15:00  59089.0  72900.0     0.0    -4.3     5.9     1.1     7.4      8.4    126.5
2020-08-28 20:16:00  59089.0  72960.0     0.0    -4.6     5.7     0.6     7.4      5.0    128.8
2020-08-28 20:17:00  59089.0  73020.0     0.0    -4.1     6.0    -0.4     7.2     -3.0    124.1
2020-08-28 20:20:00  59089.0  73200.0     0.0    -4.0     3.3    -5.0     7.2    -44.1    140.2
2020-08-28 20:21:00  59089.0  73260.0     0.0    -4.2     3.5    -5.2     7.5    -43.4    140.8

[113 rows x 9 columns]
```
which is pretty cool.

**Comment by rstoneback** [comment]
> ```
> In [9]: ace = pysat.Instrument('sw', 'ace', tag='realtime', sat_id='epam')                                                                                                             
> pysat WARNING: Unable to find any files that match the supplied template. If you have the necessary files please check pysat settings and file locations (e.g. pysat.pysat_dir).
> 
> In [10]: ace.download(dt.datetime(2001, 1, 1), dt.datetime(2002, 1, 1))                                                                                                                
> WARNING:root:real-time data only available for current day
> ```
> 
> 
> pysat is downloading data for the time range specified. The warning shouldn't be present.

Took me awhile to realize what was going on. Files were labeled with an old date but the content was for 'today'. I think the warning should be an error. Users could end up where I did, an ACE directory with a number files with filename dates that aren't valid.

Am I also missing something basic on the download calls for historic data that didn't work for me?

File: pysatSpaceWeather/instruments/sw_ace.py [hunk 4]
```diff
@@ -0,0 +1,463 @@
...
+    ----
+    Called by pysat. Not intended for direct use by user.
+
+    """
+
+    # Save each file to the output DataFrame
+    data = pds.DataFrame()
+    for fname in fnames:
+        result = pds.read_csv(fname, index_col=0, parse_dates=True)
+        data = pds.concat([data, result], axis=0)
```
Comment (rstoneback) [inline pysatSpaceWeather/instruments/sw_ace.py:120]:
```suggestion
    data = []
    for fname in fnames:
        result = pds.read_csv(fname, index_col=0, parse_dates=True)
        data.append(result) 
    data = pds.concat(data, axis=0, sort=True)
```

**Pull request merged**